server/audit_math/suite.py

```diff
@@ -1,718 +0,0 @@
-
-        Outputs:
-            pvalue - the pvalue for the hypothesis given the null_lambda
-        """
-
-        if self.sample_size == 0 or reported_margin == 0:
-            return 1.0
-
-        if self.sample_size == self.num_ballots:
-            return 0.0
-
-        o1, o2, u1, u2 = (
-            self.misstatements[(winner, loser)]["o1"],
-            self.misstatements[(winner, loser)]["o2"],
-            self.misstatements[(winner, loser)]["u1"],
-            self.misstatements[(winner, loser)]["u2"],
-        )
-
-        U_s = Decimal(2 * self.num_ballots / reported_margin)
-        gamma = Decimal(GAMMA)
-        multiplier = 1 - Decimal(null_lambda) / (gamma * U_s)
-
-        # This represents an invalid alternative, because lambda is too big.
-        if multiplier <= 0:
-            raise ValueError("Alternative hypothesis is invalid!")
-
-        log_pvalue = (
-            self.sample_size * multiplier.ln()
-            - o1 * (1 - 1 / (2 * gamma)).ln()
-            - o2 * (1 - 1 / gamma).ln()
-            - u1 * (1 + 1 / (2 * gamma)).ln()
-            - u2 * (1 + 1 / gamma).ln()
-        )
-        pvalue = (log_pvalue).exp()
-        return float(np.min([float(pvalue), 1.0]))  # cast for the typechecker
-
-
-def maximize_fisher_combined_pvalue(
-    alpha: float,
-    contest: Contest,
-    bp_stratum: BallotPollingStratum,
-    cvr_stratum: BallotComparisonStratum,
-    winner: str,
-    loser: str,
-    stepsize: float = 0.05,
-) -> float:
-    """
-    Grid search to find the maximum P-value.
-    Find the smallest Fisher's combined statistic for P-values obtained
-    by testing two null hypotheses at level alpha using data X=(X1, X2).
-
-    This was largely taken from CORLA18.
-
-    Inputs:
-        alpha: the risk limit as a fraction
-        contest: the contest being audited
-        bp_stratum: a stratum object containing the ballot polling stratum
-        cvr_stratum: a stratum object containing the ballot comparison stratum
-        winner, loser: the winner-loser pair to maximuze the p-value over
-
-    Outputs:
-        a maximum p-value for the Fisher's combining function given the two
-        strata.
-    """
-    maximized_pvalue = 0.0
-    # find range of possible lambda
-    cvr_winner_votes = cvr_stratum.vote_totals[winner]
-    bp_winner_votes = bp_stratum.vote_totals[winner]
-
-    cvr_loser_votes = cvr_stratum.vote_totals[loser]
-    bp_loser_votes = bp_stratum.vote_totals[loser]
-
-    V = cvr_winner_votes - cvr_loser_votes + bp_winner_votes - bp_loser_votes
-
-    # The election is tied
-    if V == 0:
-        return 1.0
-
-    lambda_lower = (
-        np.amax(
-            [
-                cvr_winner_votes - cvr_loser_votes - cvr_stratum.num_ballots,
-                V - (bp_winner_votes - bp_loser_votes + bp_stratum.num_ballots),
-            ]
-        )
-        / V
-    )
-    lambda_upper = (
-        np.amax(
```

Comment:
this should be `np.amin`

Reply:
it feels like there's a broader discussion of why this is a max/min at all. It seems to me like the first value is the right one per the paper. The other value is useful if there are a lot of no-vote ballots in the CVR stratum I guess? In which case max seems right to me?

Reply:
Hmm, in section 2.2 of the paper, it says:
> Some values of λ can be ruled out a priori, because (for instance) ωwℓ,s ≤ Vwℓ,s + Ns, where Ns is the number of ballots cast in stratum s, and thus 1 − (Vwℓ2 + N2) / Vwℓ ≤ λ ≤ (Vwℓ,1 + N1) / Vwℓ

Due to the "for instance," I read that as a non-exhaustive exploration of bounds - i.e. there could be other bounds as well.

In the comment for the function that the code here is based on, there's a derivation that gives the bounds implemented in the code. Seems reasonable to me to trust that.

My guess as to why they use `min` for the upper bound is that if both bounds are valid, whichever one is smaller is stricter and therefore more useful.

Reply:
aha, that makes total sense. Will fix.